####This code demonstrates how the main and supplementary figures were generated. 
###Most major computational steps are performed elsewhere, as pointed to by the "load" commands throughtout the code

setwd("") #Set to the folder where all scripts and files are found
lapply(list("pheatmap","monocle","dplyr","Seurat","parallel","doParallel","ggplot2","magrittr","reshape2","scales","splines","zoo","RColorBrewer","reldist","entropy","grid","gridExtra","plotrix","stringr","png","igraph","scatterpie"),require,character.only = TRUE)
source("InternalFunctions.R") #load our DropSeq Functions
set.seed(123)
GeneInfo<-read.csv("GeneInfoTable_expanded_May2017_PlasmoDB32.csv",stringsAsFactors = F);
ggcols<-c(gg_color_hue(16)[1:11], "magenta1","deeppink1","deeppink2","hotpink","pink")
load("ap2g-dd_seurat.RData") # see AllAP2Gonly_treat,UMI_regressed_Setup.R
Figures<-list()

############																FIGURE 1  																	 ################## 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #

###############   Figure 1B: Violin Plots
load("Figure1_seuratObject.RData") #generated by "figure1_seurat.R"
plotme<-cbind(PfDS@data.info,PfDS@tsne.rot) # <--- FOR GGPLOT

Figures[["fig1C"]]<-ggplot(plotme,aes(x=tSNE_1,y=tSNE_2,color=sample))+geom_point(size=0.5)+coord_fixed(ratio = 0.8)+
  scale_color_manual(values = c("dodgerblue4","salmon","firebrick4","skyblue","forestgreen","palegreen2","darkviolet"))+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),legend.position = "none")+
Figures[["fig1C"]]

############################################################################################################
############																FIGURE 2  																	 ################## 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #
# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv #
load("ap2g-dd_seurat.RData") #see "Seurat_setup_code.R"
ggcols<-c(gg_color_hue(16)[1:11], "magenta1","deeppink1","deeppink2","hotpink","pink")

plotme<-cbind(
  PfDS@data.info,PfDS@tsne.rot,
  exprAP2G=(exp(PfDS@data["PF3D7_1222600",])-1)
)
plotme %>% dplyr::group_by(cluster) %>% dplyr::summarize(x = median(tSNE_1), y = median(tSNE_2)) -> centers

###############   Figure 2A left: tSNE plot with cells colored by cluster, labeled

Figures[["fig2A.left"]]=ggplot(plotme,aes(x=tSNE_1,y=tSNE_2,fill=cluster,color=cluster,group=cluster))+
  coord_fixed(ratio = 1)+scale_fill_manual(values=ggcols)+scale_color_manual(values=ggcols)+
  geom_point(size=0.5,alpha=0.8,shape=21)+geom_text(data=data.frame(centers),aes(x=x,y=y,label=cluster),size=8,color="black")+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        legend.position="none"
  )
Figures[["fig2A.left"]]

###############   Figure 2A top: tSNE plot with cells colored by tp collected
plotme$tp=factor(plotme$tp, labels = c("30 hpi","36 hpi","42 hpi","Stage I GC"))
Figures[["fig2A.top"]]=ggplot(plotme,aes(x=tSNE_1,y=tSNE_2,fill=cluster,color=tp))+
  geom_point(shape=19,size=0.3,alpha=0.3)+coord_fixed(ratio = 1.0)+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        legend.position="bottom",legend.title=element_text(size=11),legend.text = element_text(size=11))+
  guides(fill=F, colour = guide_legend(override.aes = list(size=4, alpha=1.0),title="",nrow=2,byrow=TRUE))
Figures[["fig2A.top"]]

###############   Figure 2A bottom: tSNE plot of tp composition pie charts ####Needs fixing as of June 19th

plotme2<-as.data.frame.matrix(table(plotme$cluster,plotme$tp))
plotme2<-apply(plotme2,2,function(x)x/sum(x)) #scaled by ON/OFF
plotme2<-t(apply(plotme2,1,function(x)x/sum(x))) # scaled to 1 within each cluster

plotme3<-as.data.frame.matrix(table(plotme$cluster,plotme$tp))
plotme3<-as.data.frame(apply(plotme3,1,sum))
plotme3$cluster<-rownames(plotme3)
colnames(plotme3)<-c("cells","cluster")
plotme3$radius<-sqrt(plotme3$cells/pi)

tSNE_1 <- centers$x; tSNE_2 <- centers$y
d <- data.frame(tSNE_1,tSNE_2)
d[1,1]=0.225
d[2,1]=0.07
d[5,1]=-0.28;d[5,2]=-0.09
d[14,2]=0.1
d[15,2]=-0.125
d[16,1]=-0.1
d$cluster<-1:16
d<-cbind(d,plotme2)
d$radius<-plotme3$radius
Figures[["fig2A.bottom"]] <- ggplot(d,aes(x=tSNE_1,y=tSNE_2,fill=cluster,label=cluster,group=cluster))+coord_fixed(ratio = 1)+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(),axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank())+xlab("")+ylab("")+ 
  geom_scatterpie(aes(x=tSNE_1, y=tSNE_2, group=cluster, r=radius*0.004),data=d, cols=c("30 hpi","36 hpi","42 hpi","Stage I GC"), alpha=.8)+theme(legend.position = "none")
Figures[["fig2A.bottom "]]

###############   Fig. 2B		Heatmap plot with cells colored by assigned RNAseq TP

PfDS.cor.all.reorder=data.frame(PfDS.cor.all.rs[order(PfDS@data.info$cluster),]); 

internalClusterOrder=NULL
for (i in sort(unique(PfDS@data.info$cluster))){
	x=apply(PfDS.cor.all.reorder[PfDS@data.info$cluster[order(PfDS@data.info$cluster)]==i,],1,which.max)
	internalClusterOrder=c(internalClusterOrder,length(internalClusterOrder)+order(x))
}
PfDS.cor.all.reorder2=data.frame(PfDS.cor.all.reorder[internalClusterOrder,]); 
rownames(PfDS.cor.all.reorder2)=rownames(PfDS@data.info)[order(PfDS@data.info$cluster)][internalClusterOrder]
colnames(PfDS.cor.all.reorder2)=c("7-15 hpi","12-20 hpi","17-25 hpi","22-30 hpi","27-35 hpi","32-40 hpi","40-5 hpi","2-10 hpi")
annotation_row=data.frame(row.names = rownames(PfDS.cor.all.reorder2),cluster=PfDS@data.info$cluster[order(PfDS@data.info$cluster)][internalClusterOrder])
ann_colors<-list(cluster=ggcols);names(ann_colors[[1]])<-1:16
pheatmap(PfDS.cor.all.reorder2[annotation_row$cluster %in% 1:11,],cluster_cols = F,cluster_rows = F,show_rownames = F,annotation_row = annotation_row, annotation_colors = ann_colors,scale="row",,annotation_legend=F,legend_breaks=c(2,1,0,-1,-2),legend_labels=c(1,0.5,0,-0.5,-1),color = colorRampPalette(c("purple", "white", "firebrick1"))(200),annotation_names_row = F)


###############   Fig. 2C		Monocle Cell trajectory (ON only)
load("./ap2g-dd_treatedAsexuals_monocle.RData") ##See "monocle_analysis.R"

HSMM@reducedDimS[2,]=HSMM@reducedDimS[2,]
HSMM@reducedDimS[1,]=HSMM@reducedDimS[1,]
plotme=data.frame(pData(HSMM),dims=t(HSMM@reducedDimS))
Figures[["fig2C"]]=ggplot(plotme,aes(x=dims.1,y=dims.2,color=cluster))+
  geom_point()+coord_fixed(ratio = 1.0)+scale_color_manual(values=ggcols[1:11])+
  xlab("Component 1")+ylab("Component 2")+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +guides(color = F)
Figures[["fig2C"]]

###############   Fig. 2D Gene Set Expression over pseudo-time
geneModules=read.table("GeneModules.tab",stringsAsFactors = F,sep = "\t",header=T)
geneModules=geneModules[!(geneModules$Group %in% c("RH","rhoptry neck","MSP")),]
filtByExp=apply(sapply(1:11,function(y) rowMeans(subset@data[as.character(geneModules$X.Gene.ID.),subset@ident==y]>0)),1,function(x) max(x)>0)
geneModules=geneModules[filtByExp,]
expByModule=plyr::ddply(geneModules,plyr::.(Group),.fun=function(df) colSums(matrix(exp(as.matrix(subset@data[df$X.Gene.ID.,]))-1,ncol = ncol(subset@data))))
rownames(expByModule)=expByModule$Group; expByModule$Group=NULL
expByModule=data.frame(pt=subset@data.info[,c("pt")],(t(expByModule)))
expByModule=data.frame(pt=expByModule$pt,t(t(expByModule[,2:ncol(expByModule)])/apply(expByModule[,2:ncol(expByModule)],2,mean))) 
expByModule=melt(expByModule,id.vars = c("pt"));
expByModule$variable=factor(expByModule$variable,level=c("DNA_Rep.","SERA","IMC","rhoptry","microneme","CDPK5")) 

df=subset@data.info[,c("pt","cluster")]
df=df[order(df$pt),]
windownumber=100
windowsize=max(df$pt)/(windownumber/2)
fun<-function(df,low,high){mean(as.numeric(df[df$pt>low & df$pt<=high,"cluster"]))}

window<-data.frame() ; 
for(i in 1:windownumber){
  window<-rbind(window,c(i=i,pt=i*windowsize/2-windowsize/2,cluster=fun(df,(i-1)*(windowsize/2),i/2*windowsize)))
  }
colnames(window)=c("i","pt","cluster")

Figures[["fig2D"]]<-ggplot()+
  geom_rect(data=window,aes(xmin=pt-windowsize/2,xmax=pt+windowsize/2,ymin=-0.7,ymax=-0.2,fill=cluster))+scale_fill_gradientn(colours = ggcols[1:11])+
  stat_smooth(data = expByModule,aes(x=pt,y=value,color=variable),method = "loess",span = 0.3,se = F)+
  scale_y_continuous(limits=c(-0.7,NA),breaks=0:10)+
  scale_x_continuous(breaks=seq(0,60,10))+xlab("pseudo-time")+ylab("gene set expression")+guides(fill="none",color="none")+
  theme(legend.title = element_blank(),legend.position = c(0.2,0.8))
Figures[["fig2D"]]

############################################################################################################
############																FIGURE 3  																	 ################## 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #
# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv #
load("ap2g-dd_seurat.RData") #see "Seurat_setup_code.R"
ggcols<-c(gg_color_hue(16)[1:11], "magenta1","deeppink1","deeppink2","hotpink","pink")

plotme<-cbind(
	PfDS@data.info,PfDS@tsne.rot,
	exprAP2G=(exp(PfDS@data["PF3D7_1222600",])-1)
)	
plotme %>% dplyr::group_by(cluster) %>% dplyr::summarize(x = median(tSNE_1), y = median(tSNE_2)) -> centers

###############   Figure 3A: Tsne plot of Cluster 1-11 with Ap2-G expression

plotme2<-plotme[plotme$tp!="GC2" & plotme$cluster %in% 1:11,]
table(plotme2$exprAP2G>0,plotme2$treat)

Figures[["fig3A"]]=ggplot(plotme2[plotme2$treat=="ON",],aes(x=tSNE_1,y=tSNE_2,fill=exprAP2G>0,group=cluster))+
  geom_point(shape=21,alpha=0.5,aes(size=exprAP2G>0))+scale_size_manual(values = c(1,2))+
  scale_fill_manual(values=c("white","firebrick1"),labels=c("Negative","Positive"))+
  coord_fixed(ratio = 1)+
  guides(size="none",fill="none")+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        panel.background = element_rect(fill = "transparent",colour = NA),plot.background = element_rect(fill = "transparent",colour = NA))
Figures[["fig3A"]]


###############   Figure 3: AP2g % and expression by cluster Bar plots

plotme2<-plyr::ddply(plotme[plotme$tp!="GC2"&plotme$cluster%in% 1:11,],plyr::.(cluster,treat),summarize,mean=mean(exprAP2G>0)*100,cellsPos=sum(exprAP2G>0),cellNum=length(exprAP2G))

Figures[["fig3B.top"]]<-ggplot(plotme[plotme$tp!="GC2"&plotme$cluster %in% 1:11 & plotme$strain=="AP2G",], aes(x=cluster,y=exprAP2G,fill=cluster,group=treat,alpha=treat))+
  stat_summary(fun.y = mean, geom = "bar",position="dodge",color="black")+
  stat_summary(fun.data = mean_se, geom = "errorbar",position="dodge",alpha=1)+
  scale_fill_manual(values=ggcols)+
	scale_alpha_discrete(range=c(0.25,1),labels=c("untreated","treated"))+scale_y_continuous(breaks=seq(0,25,5))+
	theme(legend.justification=c(0.05,0.9), legend.position=c(0.05,0.9),legend.title=element_blank(), axis.title.y = element_text(size=10),axis.text.x = element_blank())+
  ylab("AP2-G transcripts per 10,000")+xlab("")+guides(fill=FALSE,group=F,alpha=F)
Figures[["fig3B.top"]]

Figures[["fig3B.bottom"]]<-ggplot(plotme2,aes(x=cluster,y=mean,fill=cluster,alpha=treat, label=cellNum))+
	geom_bar(stat="identity",position = "dodge",color="black")+
	scale_alpha_discrete(range=c(0.25,1),labels=c("untreated","treated"))+
	scale_fill_manual(values=ggcols)+scale_color_manual(values=ggcols)+
	ylab("AP2-G positive cells \n(%)")+xlab("Cluster")+theme(legend.position="none",axis.title.y = element_text(size=10))+
	annotate(geom = "text", x=seq(0.7,11.2,by=0.5),y=plotme2$mean+7, label = plotme2$cellNum,angle = 90)+
  annotate("text", x= rep(1:11,each=2)+rep(c(-0.25,0.25),times=11), y= -4, label=rep(c("U","T"),11),size = 4)+
  scale_y_continuous(breaks = seq(0,40,10), limits = c(-5,45))+guides(fill=FALSE)
Figures[["fig3B.bottom"]]

##Calculate significance of differences in ap2g expression
plyr::dlply(plotme[plotme$tp!="GC2" & plotme$cluster %in% 1:11,],plyr::.(cluster),.fun=function(x){11*t.test(x$exprAP2G[x$treat=="OFF"],x$exprAP2G[x$treat=="ON"])$p.value})
p.adjust(unlist(plyr::dlply(plotme[plotme$tp!="GC2" & plotme$cluster %in% 1:11,],plyr::.(cluster),.fun=function(x){t.test(x$exprAP2G[x$treat=="OFF"],x$exprAP2G[x$treat=="ON"])$p.value})),method = "fdr")

###############   Figure 3D: Heatmap of AP2-G differntially expressed genes Heatmap of the shared hits from NF54 and AP2Gdd

load("DE_EachStrainSeparately.RData") #Generated using "DifferentialExP_ByStrain.R"
###Create a seurat object combining both NF54 and AP2Gdd cells (by modifiy "Seurat_setup_code.R")
PfDS@data.info$treat[PfDS@data.info$strain=="NF54"]="ON"
PfDS@data.info$strain.tp=paste(PfDS@data.info$strain,PfDS@data.info$tp,sep="_")
PfDS@data.info$ap2g.treat=paste(PfDS@data["PF3D7_1222600",]>0,PfDS@data.info$treat,sep="_")
subset=SubsetData(PfDS,cells.use=rownames(PfDS@data.info[PfDS@data.info$treat=="ON",]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[subset@data.info$tp!="GC2",]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[!(subset@data.info$tp==36 & subset@data.info$strain=="NF54"),]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[subset@data.info$cluster %in% 1:11,]))

hitsFunction=data.frame(row.names=c("PF3D7_1238500 protein","PArt","PF3D7_0617200 protein","PF3D7_0317700 protein","PF3D7_1133700 protein","PF3D7_0406500 protein","ApiAP2 PF3D7_1139300","ApiAP2 PF3D7_1222400","ISWI","LSD2","HDA1","SNF2L","RH6","MSP11","SURF4.1","SURF4.2","MSRP1","SERA4","MSP9","SURF8.2"),
                        category=c(rep("U",3),rep("R",3),rep("D",2),rep("N",4),rep("H",8)))

fun<-function(cluster,SO,genes,ths=10){
  ONInClusAP2G<-rownames(SO@data.info[SO@data.info$ap2g.treat=="TRUE_ON" & SO@data.info$cluster==cluster & SO@data.info$strain=="AP2G",])
  OFFInClusAP2G<-rownames(SO@data.info[SO@data.info$ap2g.treat=="FALSE_ON"  & SO@data.info$cluster==cluster & SO@data.info$strain=="AP2G",])
  avg_diffAP2G<-round(log2(rowMeans(exp(as.matrix(SO@data[genes,ONInClusAP2G]))-1))-log2(rowMeans(exp(as.matrix(SO@data[genes,OFFInClusAP2G]))-1)),3)
  ONInClusNF54<-rownames(SO@data.info[SO@data.info$ap2g.treat=="TRUE_ON" & SO@data.info$cluster==cluster & SO@data.info$strain=="NF54",])
  OFFInClusNF54<-rownames(SO@data.info[SO@data.info$ap2g.treat=="FALSE_ON"  & SO@data.info$cluster==cluster & SO@data.info$strain=="NF54",])
  avg_diffNF54<-round(log2(rowMeans(exp(as.matrix(SO@data[genes,ONInClusNF54]))-1))-log2(rowMeans(exp(as.matrix(SO@data[genes,OFFInClusNF54]))-1)),3)
  badAP2G=which(rowSums(as.matrix(SO@data[genes,ONInClusAP2G])!=0)<(ths*length(ONInClusAP2G)/100) & rowSums(as.matrix(SO@data[genes,OFFInClusAP2G]!=0))<(ths*length(OFFInClusAP2G)/100))
  badNF54=which(rowSums(as.matrix(SO@data[genes,ONInClusNF54]!=0))<ths & rowSums(as.matrix(SO@data[genes,OFFInClusNF54]!=0))<ths)
  avg_diffAP2G[badAP2G]=NA;avg_diffNF54[badNF54]=NA
  return(data.frame(row.names=NULL,geneID=genes,cluster=rep(cluster,length(genes)),avg_diffAP2G,avg_diffNF54))
}

##Getting the hits repeating in both strains
top=intersect(deGenesAllClusNF54$geneID,deGenesAllClusAP2G$geneID)

hits=data.frame()
for (i in 1:11){
  hits=rbind(hits,fun(i,subset,top,ths=5))
}
heatable=reshape(hits, idvar = "geneID", timevar = "cluster",direction = "wide");
heatable=heatable[,c("geneID","avg_diffAP2G.1","avg_diffAP2G.2","avg_diffAP2G.3","avg_diffAP2G.4","avg_diffAP2G.5","avg_diffAP2G.6","avg_diffAP2G.7","avg_diffAP2G.8","avg_diffAP2G.9","avg_diffAP2G.10","avg_diffAP2G.11","avg_diffNF54.1","avg_diffNF54.2","avg_diffNF54.3","avg_diffNF54.4","avg_diffNF54.5","avg_diffNF54.6","avg_diffNF54.7","avg_diffNF54.8","avg_diffNF54.9","avg_diffNF54.10","avg_diffNF54.11")]

annotation_row<-data.frame(row.names = as.character(getInfo(heatable$geneID,cols="name")),motifs=as.character(getInfo(heatable$geneID,cols="ap2gMotifs_3k")))
annotation_col<-data.frame(row.names = paste(rep(c("AP2Gcluster","NF54cluster"),each=11),rep(1:11,2),sep = "."),cluster=as.character(rep(1:11,2)),strain=rep(c("AP2G","NF54"),each=11))
ann_colors<-list(motifs=c( colorRampPalette(c("white", "darkorchid4"))(8)),cluster=ggcols[1:11]);

names(ann_colors[["motifs"]])<-names(table(annotation_row$motifs));names(ann_colors[["cluster"]])<-1:11
rownames(heatable)<-gsub(" protein","",as.character(GeneInfo[match(heatable$geneID,GeneInfo$geneID),]$altName)); 

colnames(heatable)<-c("geneID",paste(rep(c("AP2Gcluster","NF54cluster"),each=11),rep(1:11,2),sep = "."))
heatable[is.na(heatable)]=NA
rangeFinite=range(sort(as.matrix(heatable[,-1]))[which(!is.infinite(sort(as.matrix(heatable[,-1]))))]) #range of non infinite values

heatable[,-1][heatable[,-1]>5]=5
heatable[,-1][heatable[,-1]< -3]=-3

paletteLength <- 50
myColor <- colorRampPalette(c("purple", "gray92", "firebrick1"))(paletteLength)
myBreaks <- c(seq(-5, 0, length.out=ceiling(paletteLength/2) + 1),
              seq(5/paletteLength, 5, length.out=floor(paletteLength/2)))

pheatmap(heatable[order(unlist(apply(heatable[,-1],1,which.max))),-1],cluster_cols = F,cluster_rows = F,show_rownames = T,cex=.9,annotation_row = annotation_row, annotation_col = annotation_col, annotation_colors=ann_colors, border_color = NA,treeheight_row=20,color = colorRampPalette(c("blue", "white", "red"))(50),breaks = myBreaks,fontsize_col=14,fontsize_row=14,show_colnames=F,annotation_names_col=F,annotation_legend = F,annotation_names_row = F,legend_breaks = seq(-8,8,2),gaps_col = 11,na_col = "gray85")
ftable(table(subset@data["PF3D7_1222600",]>0,subset@data.info$cluster,subset@data.info$strain))

#p.values for enrichment in motifs
motifsInOutHeatm=cbind(table("in"=factor(annotation_row$motifs,levels = 1:64)),table("out"=factor(GeneInfo$ap2gMotifs_3k[!(GeneInfo$name %in% rownames(annotation_row))],levels = 1:64)))
motifsInOutHeatm_means=c(mean(as.numeric(annotation_row$motifs))-1,mean(GeneInfo$ap2gMotifs_3k[!(GeneInfo$name %in% rownames(annotation_row))],na.rm = T))
enrichmentFold=motifsInOutHeatm_means[1]/motifsInOutHeatm_means[2]
motifsInOutHeatm=motifsInOutHeatm[rowSums(motifsInOutHeatm)>0,]
fisher.test(motifsInOutHeatm,workspace = 1e6)

#Correlation between fold-change number of motifs
cor(as.numeric(as.character(annotation_row$motifs)),apply(heatable[,-1],1,function(x) max(x,na.rm = T)),method = "spearman")
table(subset@data.info$ap2g,subset@data.info$cluster)

############																FIGURE 4  																	 ################## 
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #
# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv #
###############   Figure 4A: Expression of AP2G in NF54, AP2G ON/OFF by Pseudotime
###Create a seurat object combining both NF54 and AP2Gdd cells (by modifiy "Seurat_setup_code.R")
###Create a monocle object combining both NF54 and AP2Gdd cells (by modifiy "Seurat_setup_code.R")

set.seed(10027)
subset=SubsetData(PfDS, cells.use = rownames(PfDS@data.info[PfDS@data.info$tp!="GC2",]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[subset@data.info$cluster %in% 1:11,]))
subset@data.info$pt=pData(HSMM)$Pseudotime
subset@data.info$ap2gPos=subset@data["PF3D7_1222600",]>0

gene="PF3D7_1222600"
geneExp=10^4*subset@raw.data[gene,names(subset@ident)]/colSums(subset@raw.data[,names(subset@ident)])
expByModule=data.frame(subset@data.info[,c("pt","treat","ap2gPos","strain")],geneEx=geneExp); rm(geneExp)
expByModule=melt(expByModule,id.vars = c("treat","pt","ap2gPos","strain")); 
expByModule$variable=gsub("geneEx.","",expByModule$variable)
df=subset@data.info[,c("pt","cluster")]
df=df[order(df$pt),]
windownumber=100
windowsize=max(df$pt)/(windownumber/2)
fun<-function(df,low,high){mean(as.numeric(df[df$pt>low & df$pt<=high,"cluster"]))}

window<-data.frame() ; 
for(i in 1:windownumber){
  window<-rbind(window,c(i=i, pt=i*windowsize/2-windowsize/2,cluster=fun(df,(i-1)*(windowsize/2),i/2*windowsize)))
}
colnames(window)=c("i","pt","cluster")

#FOR NF54 alone
# indNan=is.na(window$cluster); 
# for (i in which(indNan)){
#   window$cluster[i]=mean(window$cluster[i-1],window$cluster[i+1])
# }

expByModule$treat=plyr::revalue(expByModule$treat, c("Ind"="NF54", "ON"="AP2-G-DD treated","OFF"="AP2-G-DD untreated"))
Figures[["fig4A"]]<-ggplot()+
  geom_rect(data=window,aes(xmin=pt-windowsize/2,xmax=pt+windowsize/2,ymin=-2.7,ymax=-0.2,fill=cluster))+scale_fill_gradientn(colours = ggcols[1:11])+
  stat_smooth(data = expByModule,aes(x=pt,y=value,linetype=treat,color=treat),method = "loess",span = 0.3,se = F)+ #for ap2g
  scale_linetype_manual(values=c("solid","solid","solid"))+
  scale_color_manual(values = c("deepskyblue","grey40","darkorange")) +
  scale_y_continuous(limits=c(-2.7,NA),breaks=seq(0,30,10))+
  scale_x_continuous(breaks=seq(0,60,10))+xlab("pseudo-time")+ylab(paste(getInfo(gene,"name"),"expression"))+guides(fill="none")+
  theme(legend.title = element_blank(),legend.position = c(0.35,0.8),legend.key.width=unit(1.2,"cm"),legend.text = element_text(size=10),axis.text.y = element_text(size=9),axis.title.y = element_text(size=14))#+ggtitle(gename)
Figures[["fig4A"]]

###### Gene by treatment by pt - figure 4B
mainFigureGenes=c("PF3D7_1104200","PF3D7_0624600","PF3D7_1222400","PF3D7_1472200","PF3D7_0801900","PF3D7_1139300")
geneExp=10^4*subset@raw.data[mainFigureGenes,names(subset@ident)]/colSums(subset@raw.data[,names(subset@ident)])
expByModule=data.frame(subset@data.info[,c("pt","treat","ap2gPos","strain")],geneEx=t(geneExp)); rm(geneExp)
colnames(expByModule)[5:ncol(expByModule)]=gsub(" protein","",as.character(getInfo(gsub("geneEx.","",colnames(expByModule)[5:ncol(expByModule)]),"name")))
expByModule=melt(expByModule,id.vars = c("treat","pt","ap2gPos","strain")); 
expByModule[expByModule$treat=="Ind","treat"]="ON"
expByModule$ap2gPos=plyr::revalue(as.character(expByModule$ap2gPos), c("TRUE"="+", "FALSE"="-"))
expByModule$group=paste0(expByModule$strain,expByModule$ap2gPos,expByModule$treat); expByModule$group[grep("OFF",expByModule$group)]="OFF"
expByModule$group=factor(expByModule$group, levels=c("OFF","AP2G-ON","AP2G+ON","NF54-ON","NF54+ON"))

Figures[["fig4B"]]<-
  ggplot(data = expByModule,aes(x=pt,y=value,linetype=group,color=group))+stat_smooth(method = "loess",span = 0.6,se = F,size=0.7)+
  scale_color_manual(values=c("grey40","deepskyblue","deepskyblue","darkorange","darkorange"))+scale_linetype_manual(values=c("solid","dashed","solid","dashed","solid"))+
  scale_y_continuous(limits=c(-0.7,NA))+
  scale_x_continuous(breaks=seq(0,60,10))+xlab("pseudo-time")+ylab("transripts per 10,000")+guides(linetype=F,color=F)+
  theme(axis.text.y = element_text(),axis.title.y = element_text(),strip.text = element_blank(),strip.background = element_blank())+
  facet_wrap(~variable,ncol = 3,scales = "free")
Figures[["fig4B"]]

###############   Figure 4E: Co-expression networks in AP2G+nf54 
###Create a seurat object combining both NF54 and AP2Gdd cells (by modifiy "Seurat_setup_code.R")
PfDS@data.info$ap2gPos=PfDS@data["PF3D7_1222600",]>0
subset=SubsetData(PfDS,cells.use=rownames(PfDS@data.info[PfDS@data.info$treat!="OFF",]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[subset@data.info$tp!="GC2",]))
subset=SubsetData(subset, cells.use = rownames(subset@data.info[subset@data.info$cluster %in% 1:11,]))
PfDS=subset;rm(subset)
load("DE_EachStrainSeparately.RData") #Generated using "DifferentialExP_ByStrain.R"
hits=c("PF3D7_1222600",intersect(deGenesAllClusNF54$geneID,deGenesAllClusAP2G$geneID))#nf54+ap2g alone
AP2GbyClusterTreat.coX.spear<-multicoX(geneListA=hits, geneListB="all", exprMat = PfDS@data, by=PfDS@data.info$cluster,parallel=T,corMeth="spearman")

AP2GbyClusterTreat.coX.spear<-AP2GbyClusterTreat.coX.spear[AP2GbyClusterTreat.coX.spear$p.value<0.05,] # significant
AP2GbyClusterTreat.coX.spear<-AP2GbyClusterTreat.coX.spear[sign(AP2GbyClusterTreat.coX.spear$phi)==sign(AP2GbyClusterTreat.coX.spear$cor),] # cor & phi agree on direction
AP2GbyClusterTreat.coX.spear<-AP2GbyClusterTreat.coX.spear[AP2GbyClusterTreat.coX.spear$geneB %in% GeneInfo$geneID,]
AP2GbyClusterTreat.coX.spear$by2=levels(factor(PfDS@data.info$cluster))[AP2GbyClusterTreat.coX.spear$by]
AP2GbyClusterTreat.coX.spear$clus=AP2GbyClusterTreat.coX.spear$by
AP2GbyClusterTreat.coX.spear<-AP2GbyClusterTreat.coX.spear[AP2GbyClusterTreat.coX.spear$phi>0.3,] # significant
AP2GbyClusterTreat.coX.spear$geneAname=getInfo(AP2GbyClusterTreat.coX.spear$geneA,"name")
AP2GbyClusterTreat.coX.spear$geneBname=getInfo(AP2GbyClusterTreat.coX.spear$geneB,"name")
tab=table(AP2GbyClusterTreat.coX.spear$geneBname)
tab=tab[tab>5]
AP2GbyClusterTreat.coX.spearFilt_forSupTable=AP2GbyClusterTreat.coX.spear[AP2GbyClusterTreat.coX.spear$geneBname %in% names(tab),]

#for supp table
fishTable=data.frame(t(sapply(1:nrow(AP2GbyClusterTreat.coX.spearFilt_forSupTable),function(i){
ct=table(PfDS@data[AP2GbyClusterTreat.coX.spearFilt_forSupTable$geneA[i],PfDS@ident==AP2GbyClusterTreat.coX.spearFilt_forSupTable$by[i]]>0,
         PfDS@data[AP2GbyClusterTreat.coX.spearFilt_forSupTable$geneB[i],PfDS@ident==AP2GbyClusterTreat.coX.spearFilt_forSupTable$by[i]]>0)
ft=fisher.test(ct)
c(as.vector(ct),ft$estimate,ft$conf.int)
})))
names(fishTable)=c("Aoff_Boff","Aoff_Bon","Aon_Boff","Aon_Bon","oddsRatio2","conf1","conf2")

AP2GbyClusterTreat.coX.spearFilt_forSupTable=cbind(AP2GbyClusterTreat.coX.spearFilt_forSupTable,fishTable)

#for net
AP2GbyClusterTreat.coX.spearFilt=AP2GbyClusterTreat.coX.spear[AP2GbyClusterTreat.coX.spear$geneBname %in% names(tab),c("clus","geneA","geneB")]
AP2GbyClusterTreat.coX.spearFilt=data.frame(t(apply(AP2GbyClusterTreat.coX.spearFilt[,2:3], 1, sort)),AP2GbyClusterTreat.coX.spearFilt$clus);names(AP2GbyClusterTreat.coX.spearFilt)=c("A","B","clus")
df=plyr::count(unique(AP2GbyClusterTreat.coX.spearFilt), vars = c("A", "B"))   

g=graph.data.frame(df,directed = F)
E(g)$weight=df$freq 
V(g)$Motifs=getInfo(V(g)$name,"ap2gMotifs_3k")
V(g)$altName=getInfo(V(g)$name,"name")

AP2GbyClusterTreat.coX.spear$clus=factor(AP2GbyClusterTreat.coX.spear$clus)

########################				SUPPLEMENTARY FIGURES				################################################
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ #
#############	 Figure S3: AP2-G Cluster Analysis	 #########################
load("ap2g-dd_seurat.RData") #see "Seurat_setup_code.R"
source("InternalFunctions.R") #load our DropSeq Functions
ggcols<-c(gg_color_hue(16)[1:11], "magenta1","deeppink1","deeppink2","hotpink","pink")

plotme<-cbind(
	PfDS@data.info,PfDS@tsne.rot,
	exprAP2G=(exp(PfDS@data["PF3D7_1222600",])-1)
)
plotme %>% dplyr::group_by(cluster) %>% dplyr::summarize(x = median(tSNE_1), y = median(tSNE_2)) -> centers

######## Figure S3A
plotme2<-as.data.frame.matrix(table(plotme$tp,plotme$cluster))
plotme2<-apply(plotme2,1,function(x)x/sum(x)) #scaled by # of cells per tp
plotme2<-melt(plotme2); colnames(plotme2)<-c("cluster","tp","frequency")
plotme2$cluster<-factor(plotme2$cluster);
labels<-c("30 hpi","36 hpi","42 hpi","Stage I GCs");names(labels)<-c("30","36","42","GC2")

Figures[["figS3A"]]<-ggplot(plotme2,aes(x=1,y=frequency,fill=cluster))+geom_bar(stat="identity",color="black")+facet_wrap(~tp, ncol=2,labeller = as_labeller(labels))+coord_polar(theta="y")+
	scale_fill_manual(values=ggcols)+
	theme(legend.position="bottom",axis.ticks = element_blank(), axis.text.y = element_blank(),axis.line.x=element_blank(),axis.line.y=element_blank(),axis.text.x = element_blank(),strip.background = element_rect("white"))+
	xlab("")+ylab("")+ggtitle("Cluster Assignment by collection time point")+
  guides(fill=guide_legend(nrow=2,byrow=TRUE))
Figures[["figS3A"]]

######## Figure S3B  
plotme2<-as.data.frame.matrix(table(plotme$cluster,plotme$tp))
plotme2<-apply(plotme2,2,function(x)x/sum(x)) #scaled by # of cells per tp
plotme2<-t(apply(plotme2,1,function(x)x/sum(x))) # scaled to 1 within each cluster
plotme2<-melt(plotme2); colnames(plotme2)<-c("cluster","tp","frequency")
plotme2$tp<-factor(plotme2$tp,levels=c("GC2",42,36,30));plotme2$cluster<-factor(plotme2$cluster);
labels<-paste("cluster",1:16);names(labels)<-1:16

Figures[["figS3B"]]<-ggplot(plotme2,aes(x=1,y=frequency,fill=tp))+geom_bar(stat="identity",color="black")+
	scale_fill_manual(values=gg_color_hue(4)[4:1],labels=c("Stage I GCs",42,36,30))+facet_wrap(~cluster,labeller = as_labeller(labels))+ coord_polar(theta="y")+ 
	theme(legend.position="bottom",axis.ticks = element_blank(), axis.text.y = element_blank(),axis.text.x = element_blank(),axis.line.x=element_blank(),axis.line.y=element_blank(),strip.background = element_rect("white"))+
	xlab("")+ylab("")+ggtitle("Relative Cluster Composition by time point")+
  guides(fill=guide_legend(title="time point"))
Figures[["figS3B"]]

######## Figure S3C #fisher test for cluster enrichment
plotme3<-cbind(PfDS@data.info,PfDS@tsne.rot)
plotme3<-plotme[plotme$tp!="GC2",]
EnrichmentPvals=NULL
for (i in sort(unique(plotme3$cluster)))
{
	ft=fisher.test(table(plotme3$cluster==i,plotme3$treat)) 
	EnrichmentPvals=rbind(EnrichmentPvals,c(ft$p.value,as.numeric(ft$conf.int),as.numeric(ft$estimate)))
}
EnrichmentPvals=data.frame(EnrichmentPvals); colnames(EnrichmentPvals)=c("pval","confintLow","confintHigh","oddsRatio")
EnrichmentPvals$AdjP=p.adjust(EnrichmentPvals$pval,method = "fdr")
EnrichmentPvals$cluster=factor(1:16)

Figures[["figS3C"]]=ggplot(EnrichmentPvals, aes(x=cluster,y=log2(oddsRatio),fill=cluster))+
	geom_rect(ymin = -1,ymax=1,xmin=0.5,xmax=16.5,fill="grey90")+
	geom_bar(stat="identity",color="black")+geom_hline(yintercept=0,size=0.3)+
	geom_errorbar(aes(ymax = log2(confintHigh), ymin=log2(confintLow)), position="dodge", width=0.25)+
	scale_fill_manual(values=ggcols)+scale_y_continuous(limits=c(-2,5.2),breaks=c(-2,-1,0,1:5))+
	theme(legend.position="none")+ylab("log2 odds-ratio")+xlab("cluster")
Figures[["figS3C"]]
	
#############	 Figure S4a: Rebulked SCTs match bulk RNA-seq ###################################################	

fpkm<-read.csv("Bartfai_RNASeqTC.pct",sep="\t",stringsAsFactors = F,row.names=1)
fpkm<-fpkm[,c(2:8,1)]
shared=intersect(rownames(fpkm),rownames(PfDS@data))

meanDS<-cbind(DS30=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==30,])])),DS36=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==36,])])),DS42=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==42,])])))
pctDS<-cbind(DS30=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==30,])]>0)),DS36=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==36,])]>0)),DS42=rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==42,])]>0)))
rnkDS<-cbind(rnk30=cbind(pct30=sort(rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==30,])]>0))),rnk30=1:dim(PfDS@data[shared,])[1])[shared,-1],rnk36=cbind(pct36=sort(rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==36,])]>0))),rnk30=1:dim(PfDS@data[shared,])[1])[shared,-1],rnk42=cbind(pct42=sort(rowMeans(as.matrix(PfDS@data[shared,rownames(PfDS@data.info[PfDS@data.info$tp==42,])]>0))),rnk30=1:dim(PfDS@data[shared,])[1])[shared,-1])
rnkDS<-rnkDS/dim(rnkDS)[1]*100

plotme<-melt(cor(rnkDS[shared,],fpkm[shared,]))
plotme %>% dplyr::mutate_if(is.factor, as.character) -> plotme
plotme$Var1<-factor(rep(c("24-36h","30-42h","36-48h"),8),levels=c("36-48h","30-42h","24-36h"))
plotme$Var2<-factor(rep(c("2-10h","7-15h","12-20h","17-25h","22-30h","27-35h","27-40h","40-5h"),each=3),levels=c("2-10h","7-15h","12-20h","17-25h","22-30h","27-35h","27-40h","40-5h"))
colnames(plotme)[3]<-"Correlation"
plotme$Correlation=(round(plotme$Correlation,2))
plotme$CorrelationLab=sprintf("%.2f",plotme$Correlation)

Figures[["figS4"]]<-ggplot(plotme, aes(Var2, Var1)) +
	geom_tile(aes(fill = Correlation),color="black") + geom_text(aes(label = CorrelationLab)) +scale_fill_gradient(low = "white", high = "red") +
	ylab("Drop-seq time point")+	xlab("RNA-seq time point")
Figures[["figS4"]]

#############	 Figure S4b: Pseudotime boxplots for AP2-G-DD ###################################################	
load("./ap2g-dd_treatedAsexuals_monocle.RData") ##See "monocle_analysis.R"

df=data.frame(Pseudotime=max(pData(HSMM)$Pseudotime)-pData(HSMM)$Pseudotime,Cluster=subset@ident,"Collection Time"=subset@data.info$tp)
Figures[["figS5A"]]<-ggplot(df,aes(x=Cluster,y=Pseudotime,color=Cluster))+geom_boxplot()+scale_y_continuous(breaks=c(seq(0,60,10)))+ylab("pseudo-time")+xlab("cluster")
Figures[["figS5A"]]
Figures[["figS5B"]]<-ggplot(df,aes(x=Collection.Time,y=Pseudotime,group=Collection.Time,color=Collection.Time))+geom_boxplot()+scale_y_continuous(breaks=c(seq(0,60,10)))+guides(colour=FALSE)+xlab("collection time")+ylab("pseudo-time")
Figures[["figS5B"]]
